src/Navigation/Notifications/Depository/Notification.php

```diff
@@ -0,0 +1,141 @@
+	{
+		parent::__construct($database, $logger, $factory ?? new Factory\Notification($logger));
+	}
+
+	/**
+	 * @param array $condition
+	 * @param array $params
+	 * @return Entity\Notification
+	 * @throws NotFoundException
+	 */
+	private function selectOne(array $condition, array $params = []): Entity\Notification
+	{
+		return parent::_selectOne($condition, $params);
+	}
+
+	private function select(array $condition, array $params = []): Collection\Notifications
+	{
+		return new Collection\Notifications(parent::_select($condition, $params)->getArrayCopy());
+	}
+
+	public function countForUser($uid, array $condition, array $params = []): int
+	{
+		$condition = DBA::mergeConditions($condition, ['uid' => $uid]);
+
+		return $this->count($condition, $params);
+	}
+
+	public function existsForUser($uid, array $condition): bool
+	{
+		$condition = DBA::mergeConditions($condition, ['uid' => $uid]);
+
+		return $this->exists($condition);
+	}
+
+	/**
+	 * @param int $id
+	 * @return Entity\Notification
+	 * @throws NotFoundException
+	 */
+	public function selectOneById(int $id): Entity\Notification
+	{
+		return $this->selectOne(['id' => $id]);
+	}
+
+	public function selectOneForUser(int $uid, array $condition, array $params = []): Entity\Notification
+	{
+		$condition = DBA::mergeConditions($condition, ['uid' => $uid]);
+
+		return $this->selectOne($condition, $params);
+	}
+
+	public function selectForUser(int $uid, array $condition = [], array $params = []): Collection\Notifications
+	{
+		$condition = DBA::mergeConditions($condition, ['uid' => $uid]);
+
+		return $this->select($condition, $params);
+	}
+
+	public function selectAllForUser(int $uid): Collection\Notifications
+	{
+		return $this->selectForUser($uid);
+	}
+
+	/**
+	 * @param array    $condition
+	 * @param array    $params
+	 * @param int|null $min_id Retrieve models with an id no fewer than this, as close to it as possible
+	 * @param int|null $max_id Retrieve models with an id no greater than this, as close to it as possible
+	 * @param int      $limit
+	 * @return BaseCollection
+	 * @throws Exception
+	 * @see _selectByBoundaries
+	 */
+	public function selectByBoundaries(array $condition = [], array $params = [], int $min_id = null, int $max_id = null, int $limit = self::LIMIT)
+	{
+		$BaseCollection = parent::_selectByBoundaries($condition, $params, $min_id, $max_id, $limit);
+
+		return new Collection\Notifications($BaseCollection->getArrayCopy(), $BaseCollection->getTotalCount());
+	}
+
+	public function setAllSeenForUser(int $uid, array $condition = []): bool
+	{
+		$condition = DBA::mergeConditions($condition, ['uid' => $uid]);
+
+		return $this->db->update(self::$table_name, ['seen' => true], $condition);
+	}
+
+	/**
+	 * @param Entity\Notification $Notification
+	 * @return Entity\Notification
+	 * @throws Exception
+	 */
+	public function save(Entity\Notification $Notification): Entity\Notification
```

Comment:
Shouldn't we add an abstract `abstract public function save(BaseEntity): BaseEntity` to force Depositories having a common save method? - Or did I miss a point :X

Reply:
Unfortunately, no. Doing so prevents child `Depository` classes to declare the kind of method you commented on because PHP doesn't accept function overloading with a more precise parameter. In this case, `Entity\Notification` would be more precise than `BaseEntity` and it would trigger a parse error. This is because PHP doesn't support "generic polymorphism" if you're interested in the nitty-gritty.

This is also why `Depositories` have ugly protected `_select*` methods that child classes can only use in their own `select*` methods to be able to fully take advantage of type-hinting.

Reply:
What about using interfaces instead? AFAIK they support overloading, don't they?

Reply:
Same problem for parameter type overloading. I tried with Interfaces and it still didn't work. It does for return values which is why `ICanCreateFromTableRow` exists!